md5/src/compress/loongarch64_asm.rs

```diff
@@ -0,0 +1,206 @@
+
+use core::arch::asm;
+
+const RC: [u32; 64] = [
+    // round 1
+    0xd76aa478, 0xe8c7b756, 0x242070db, 0xc1bdceee, 0xf57c0faf, 0x4787c62a, 0xa8304613, 0xfd469501,
+    0x698098d8, 0x8b44f7af, 0xffff5bb1, 0x895cd7be, 0x6b901122, 0xfd987193, 0xa679438e, 0x49b40821,
+    // round 2
+    0xf61e2562, 0xc040b340, 0x265e5a51, 0xe9b6c7aa, 0xd62f105d, 0x02441453, 0xd8a1e681, 0xe7d3fbc8,
+    0x21e1cde6, 0xc33707d6, 0xf4d50d87, 0x455a14ed, 0xa9e3e905, 0xfcefa3f8, 0x676f02d9, 0x8d2a4c8a,
+    // round 3
+    0xfffa3942, 0x8771f681, 0x6d9d6122, 0xfde5380c, 0xa4beea44, 0x4bdecfa9, 0xf6bb4b60, 0xbebfbc70,
+    0x289b7ec6, 0xeaa127fa, 0xd4ef3085, 0x04881d05, 0xd9d4d039, 0xe6db99e5, 0x1fa27cf8, 0xc4ac5665,
+    // round 4
+    0xf4292244, 0x432aff97, 0xab9423a7, 0xfc93a039, 0x655b59c3, 0x8f0ccc92, 0xffeff47d, 0x85845dd1,
+    0x6fa87e4f, 0xfe2ce6e0, 0xa3014314, 0x4e0811a1, 0xf7537e82, 0xbd3af235, 0x2ad7d2bb, 0xeb86d391,
+];
+
+macro_rules! c {
+    ($($l:expr)*) => {
+        concat!($($l ,)*)
+    };
+}
+
+macro_rules! round0 {
+    ($a:literal, $b:literal, $c:literal, $d:literal, $k:literal, $s:literal, $i:literal) => {
+        c!(
+            "xor    $t4," $c "," $d ";"
+            "and    $t4, $t4," $b ";"
+            "xor    $t4, $t4," $d ";"
+            roundtail!($a, $b, $k, $s, $i)
+        )
+    }
+}
+
+macro_rules! round1 {
+    ($a:literal, $b:literal, $c:literal, $d:literal, $k:literal, $s:literal, $i:literal) => {
+        c!(
+            "andn    $t4," $c "," $d ";"
+            "and     $t5," $d "," $b ";"
+            "or      $t4, $t4, $t5;"
+            roundtail!($a, $b, $k, $s, $i)
+        )
+    }
+}
+
+macro_rules! round2 {
+    ($a:literal, $b:literal, $c:literal, $d:literal, $k:literal, $s:literal, $i:literal) => {
+        c!(
+            "xor    $t4," $c "," $d ";"
+            "xor    $t4, $t4," $b ";"
+            roundtail!($a, $b, $k, $s, $i)
+        )
+    }
+}
+
+macro_rules! round3 {
+    ($a:literal, $b:literal, $c:literal, $d:literal, $k:literal, $s:literal, $i:literal) => {
+        c!(
+            "orn    $t4," $b "," $d ";"
+            "xor    $t4, $t4," $c ";"
+            roundtail!($a, $b, $k, $s, $i)
+        )
+    }
+}
+
+macro_rules! roundtail {
+    ($a:literal, $b:literal, $k:literal, $s:literal, $i:literal) => {
+        c!(
+            "ld.w       $t5, $a3," $i " * 4;"
+            "ld.w       $t6, $a1," $k " * 4;"
+            "add.w      " $a "," $a ", $t5;"
+            "add.w      " $a "," $a ", $t6;"
+            "add.w      " $a "," $a ", $t4;"
+            "rotri.w    " $a "," $a ", 32 -" $s ";"
+            "add.w      " $a "," $a "," $b ";"
+        )
+    }
+}
+
+pub fn compress(state: &mut [u32; 4], blocks: &[[u8; 64]]) {
+    if blocks.is_empty() {
+        return;
+    }
+
+    unsafe {
+        asm!(
+            // Load state
+            "ld.w    $t0, $a0, 0",
+            "ld.w    $t1, $a0, 4",
+            "ld.w    $t2, $a0, 8",
+            "ld.w    $t3, $a0, 12",
```

Comment:
Instead of manually loading/storing state registers you could use:
```rust
inout("$a4") state[0],
inout("$a5") state[1],
inout("$a6") state[2],
inout("$a7") state[3],
```
It also will allow you to use `options(preserves_flags, readonly, pure, nostack)`.